framework/src/tinyalsa/tinyalsa.c

```diff
@@ -553,30 +547,77 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 		pcm->running = 1;
 	}
 
-	/* Wait for deque message from kernel */
-	size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
-	if (size != sizeof(msg)) {
-		/* Interrupted by a signal? What to do? */
-		return oops(pcm, EINTR, "Interrupted while waiting for deque message from kernel");
-	}
-	if (msg.msgId == AUDIO_MSG_DEQUEUE) {
-		apb = (struct ap_buffer_s *)msg.u.pPtr;
-		/* Copy data to user buffer */
-		memcpy(data, apb->samp, apb->nbytes);
-		/* Enque buffer for next read opertion */
-		nbytes = apb->nbytes;
-		apb->nbytes = 0;
-		apb->curbyte = 0;
-		apb->flags = 0;
-		bufdesc.u.pBuffer = apb;
-		if (ioctl(pcm->fd, AUDIOIOC_ENQUEUEBUFFER, (unsigned long)&bufdesc) < 0) {
-			return oops(pcm, errno, "failed to enque buffer after read");
+	if (pcm->nextBuf != NULL) {
+		/* We have a pending buffer with some data in it. */
+		apb = pcm->nextBuf;
+
+		if (nbytes < pcm->nextSize) {
+			/* User wants less data than we have in buffer.
+			   Copy the requested amount of data to user buffer and return */
+			memcpy(data, apb->samp + pcm->nextOffset, nbytes);
```

Comment:
Here , aren't we always wasting pcm->nextSize - nbytes of recorded data? The user may want that in a succeeding record request. I feel it is not good to discard that data. We should record less, if less amount of data is needed.